### DISCUSSION TITLE
Copy instruction (for massive, bulk load) support

### DISCUSSION THREAD

**Discussion by simonaubertbd** [sqlalchemy/sqlalchemy#12045] [Blue Sky]
### Describe the use case

Hello all,

Copy is a very commun function that allows the insertion of data in a table much faster than insert command.
Here is a benchmark on postgres
https://github.com/josethz00/copy-vs-insert

### Databases / Backends / Drivers targeted

Any database that has a copy instruction such as :
-Vertica https://docs.vertica.com/12.0.x/en/sql-reference/statements/copy/
-MonetDB https://www.monetdb.org/documentation-Aug2024/user-guide/sql-manual/data-loading/copy-from/
-Postgresql https://docs.postgresql.fr/14/sql-copy.html
-maybe Hive/impala for HDFS?

### Example Use

Something like that

COPY [ /*+ LABEL (label-string)*/ ] [[database.]schema-name.]target-table
   [ ( { column-as-expression | column }
       [ DELIMITER [ AS ] 'char' ]
       [ ENCLOSED [ BY ] 'char' ]
       [ ENFORCELENGTH ]
       [ ESCAPE [ AS ] 'char' | NO ESCAPE ]
       [ FILLER datatype]
       [ FORMAT 'format' ]
       [ NULL [ AS ] 'string' ]
       [ TRIM 'byte' ]
       [,...] ) ]
   [ COLUMN OPTION (column
       [ DELIMITER [ AS ] 'char' ]
       [ ENCLOSED [ BY ] 'char' ]
       [ ENFORCELENGTH ]
       [ ESCAPE [ AS ] 'char' | NO ESCAPE ]
       [ FORMAT 'format' ]
       [ NULL [ AS ] 'string' ]
       [ TRIM 'byte' ]
     [,...] ) ]
FROM {
   [ LOCAL ] STDIN [ [compression](https://docs.vertica.com/12.0.x/en/sql-reference/statements/copy/parameters/#compression) ]
   | { 'path-to-data'
       [ ON { nodename | (nodeset) | ANY NODE | EACH NODE } ] [ compression ] }[,...]
     [ PARTITION COLUMNS column[,...] ]
   | [LOCAL](https://docs.vertica.com/12.0.x/en/sql-reference/statements/copy-local/#) 'path-to-data' [ [compression](https://docs.vertica.com/12.0.x/en/sql-reference/statements/copy/parameters/#compression) ] [,...]
   | [VERTICA](https://docs.vertica.com/12.0.x/en/sql-reference/statements/copy-from/#) source-database.[source-schema.]source-table[( source-column[,...] ) ]
  }
  [ NATIVE
    | FIXEDWIDTH COLSIZES {( integer )[,...]}
    | NATIVE VARCHAR
    | ORC
    | PARQUET
  ]
  | [ WITH ] UDL-clause[...]
}
   [ ABORT ON ERROR ]
   [ DELIMITER [ AS ] 'char' ]
   [ ENCLOSED [ BY ] 'char'
   [ ENFORCELENGTH ]
   [ ERROR TOLERANCE ]
   [ ESCAPE [ AS ] 'char' | NO ESCAPE ]
   [ EXCEPTIONS 'path' [ ON nodename] [,...]
   [ NULL [ AS ] 'string' ]
   [ RECORD TERMINATOR 'string' ]
   [ REJECTED DATA {'path' [ ON nodename] [,...] | AS TABLE reject-table} ]
   [ REJECTMAX integer ]
   [ SKIP integer ]
   [ SKIP BYTES integer ]
   [ STREAM NAME  'streamName']
   [ TRAILING NULLCOLS ]
   [ TRIM 'byte' ]
   [ [ WITH ] PARSER parser ([ arg=value[,...] ]) ] ]
   [ NO COMMIT ]

**Comment by CaselIT** [comment]
Hi,

The main issue that I see is that each database has its own specification and set of options.
Furthermore that isn't really a standard api in the driver and each one of them may work differently.
For example in postgresql you don't really write the copy sql, but you invoke the relevant command in the driver that will do that for you: 
- https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.copy_records_to_table
- https://www.psycopg.org/psycopg3/docs/basic/copy.html
- https://www.psycopg.org/docs/usage.html#using-copy-to-and-copy-from

Also copy is usually bi-directional, meaning it can be used to write data in the db but also to read data from it.

All this to say that while a common API may be found, it doesn't look like a simple job / the resulting API may not be great.
Also, at least in postgresql, the driver interface doesn't really fit very well with the current sqlalchemy paradigm where you have a statement that you execute on a connection, since there is no statement there.

@zzzeek what's your opinion on it?

**Comment by zzzeek** [comment]
I dont think you can even use COPY for PG without using the special DBAPI-level APIs like this one: https://www.psycopg.org/psycopg3/docs/basic/copy.html    and im sure you can't have a transaction going on either

normally I'd point users to these special APIs, so no I'd not see a path to adding new APIs for this

**Comment by CaselIT** [comment]
moving to bluesky discussion, since as mike mentioned, I'm not even sure what an api would look like here
